### PR TITLE
feat(pkg): improve/simplify reorg check logic

### PR DESCRIPTION
**Pull request by YoGhurt111**
### CHANGELOG #600
1. Find a better L1 height to reset rather than the genesis height when there is a mismatch between protocol and L2 EE

### CODE REVIEW

File: driver/chain_syncer/calldata/syncer.go [hunk 1]
```diff
@@ -517,65 +522,119 @@ func (s *Syncer) createExecutionPayloads(
...
+	}
+
+	var (
+		reorgCheckResult = new(rpc.ReorgCheckResult)
+		err              error
+		currentBlockID   uint64
+		l1HeaderToSet    *types.Header
+	)
+
+	if float64(blockID) > math.Pow(2, float64(retries)) {
```
Comment (mask-pp) [inline driver/chain_syncer/calldata/syncer.go:569]:
```suggestion
	if val := uint64(2 << retries); blockID > val {
```

File: driver/chain_syncer/calldata/syncer.go [hunk 2]
```diff
@@ -517,65 +522,119 @@ func (s *Syncer) createExecutionPayloads(
...
+		currentBlockID = uint64(float64(blockID)-math.Pow(2, float64(retries))) + 1
```
Comment (mask-pp) [inline driver/chain_syncer/calldata/syncer.go:570]:
```suggestion
		currentBlockID = blockID - val + 1
```

**Review by mask-pp — APPROVED** [review]

File: cmd/flags/driver.go [hunk 3]
```diff
@@ -43,6 +43,14 @@ var (
 		Usage:    "HTTP RPC endpoint of another synced L2 execution engine node",
 		Category: driverCategory,
 	}
+	// syncer specific flag
+	MaxExponent = &cli.Uint64Flag{
```
Comment (davidtaikocha) [inline cmd/flags/driver.go:47]: Not sure if this flag is necessary, we can always simply roll-back to the latest matched verified block hash?

File: driver/chain_syncer/calldata/syncer.go [hunk 4]
```diff
@@ -517,65 +521,123 @@ func (s *Syncer) createExecutionPayloads(
...
+func (s *Syncer) checkLastVerifiedBlockMismatch(ctx context.Context) (*rpc.ReorgCheckResult, error) {
+	var (
+		reorgCheckResult = new(rpc.ReorgCheckResult)
+		err              error
+	)
+
 	stateVars, err := s.rpc.GetProtocolStateVariables(&bind.CallOpts{Context: ctx})
 	if err != nil {
-		return false, err
+		return reorgCheckResult, err
```
Comment (davidtaikocha) [inline driver/chain_syncer/calldata/syncer.go:532]:
```suggestion
		return nil, err
```
Let's use `nil` for clarity when there is an error

File: driver/chain_syncer/calldata/syncer.go [hunk 5]
```diff
@@ -517,65 +521,123 @@ func (s *Syncer) createExecutionPayloads(
...
 	if s.state.GetL2Head().Number.Uint64() < stateVars.B.LastVerifiedBlockId {
-		return false, nil
+		return reorgCheckResult, nil
 	}
 
-	blockInfo, err := s.rpc.GetL2BlockInfo(ctx, new(big.Int).SetUint64(stateVars.B.LastVerifiedBlockId))
+	reorgCheckResult, err = s.retrievePastBlock(ctx, stateVars.B.LastVerifiedBlockId, 0)
 	if err != nil {
-		return false, err
+		return reorgCheckResult, err
```
Comment (davidtaikocha) [inline driver/chain_syncer/calldata/syncer.go:541]:
```suggestion
		return nil, err
```
ditto

File: cmd/flags/driver.go (hunk 3 above)
Comment (davidtaikocha) [reply cmd/flags/driver.go:47]: And if u want to keep this flag, u also need to add this flag to the `DriverFlags` below.

File: driver/chain_syncer/calldata/syncer.go [hunk 6]
```diff
@@ -517,65 +521,123 @@ func (s *Syncer) createExecutionPayloads(
...
+		currentBlockID = blockID - val + 1
+	} else {
+		currentBlockID = 0
 	}
 
-	l2Header, err := s.rpc.L2.HeaderByNumber(ctx, new(big.Int).SetUint64(stateVars.B.LastVerifiedBlockId))
+	blockInfo, err := s.rpc.GetL2BlockInfo(ctx, new(big.Int).SetUint64(currentBlockID))
 	if err != nil {
-		return false, err
+		return reorgCheckResult, err
```
Comment (davidtaikocha) [inline driver/chain_syncer/calldata/syncer.go:576]:
```suggestion
		return nil, err
```

File: driver/chain_syncer/calldata/syncer.go [hunk 7]
```diff
@@ -517,65 +521,123 @@ func (s *Syncer) createExecutionPayloads(
...
 	}
 
-	return blockInfo.Ts.BlockHash != l2Header.Hash(), nil
+	l2Header, err := s.rpc.L2.HeaderByNumber(ctx, new(big.Int).SetUint64(currentBlockID))
+	if err != nil {
+		return reorgCheckResult, err
```
Comment (davidtaikocha) [inline driver/chain_syncer/calldata/syncer.go:581]:
```suggestion
		return nil, err
```

File: driver/chain_syncer/calldata/syncer.go [hunk 8]
```diff
@@ -517,65 +521,123 @@ func (s *Syncer) createExecutionPayloads(
...
+	if blockInfo.Ts.BlockHash == l2Header.Hash() {
+		// To reduce the number of call contracts by bringing forward the termination condition judgement
+		if retries == 0 {
+			return reorgCheckResult, nil
+		}
+		l1Origin, err := s.rpc.L2.L1OriginByID(ctx, new(big.Int).SetUint64(currentBlockID))
+		if err != nil {
+			if err.Error() == ethereum.NotFound.Error() {
+				log.Info("L1Origin not found in retrievePastBlock because the L2 EE is just synced through P2P", "blockID", blockID)
+				// cant find l1Origin in L2 EE, so we call contract to get block info
```
Comment (davidtaikocha) [inline driver/chain_syncer/calldata/syncer.go:592]:
```suggestion
				// Can't find l1Origin in L2 EE, so we call the contract to get block info
```

File: driver/chain_syncer/calldata/syncer.go [hunk 9]
```diff
@@ -517,65 +521,123 @@ func (s *Syncer) createExecutionPayloads(
...
 	if err != nil {
-		return false, err
+		return reorgCheckResult, err
+	}
+
+	return reorgCheckResult, nil
+}
+
+// retrievePastBlock find proper L1 header and L2 block id to reset when there is a mismatch
+func (s *Syncer) retrievePastBlock(ctx context.Context, blockID uint64, retries uint64) (*rpc.ReorgCheckResult, error) {
```
Comment (davidtaikocha) [inline driver/chain_syncer/calldata/syncer.go:548]: Shall we add some unit tests for this new method?

File: cmd/flags/driver.go (hunk 3 above)
Comment (YoGhurt111) [reply cmd/flags/driver.go:47]: > And if u want to keep this flag, u also need to add this flag to the `DriverFlags` below.

OK

Comment (YoGhurt111) [reply cmd/flags/driver.go:47]: > Not sure if this flag is necessary, we can always simply roll-back to the latest matched verified block hash?

If we sync the block from L2 to L3, this value will be larger than when we sync from L1 to L2, so I think it's better to keep it a configurable value

File: driver/chain_syncer/calldata/syncer.go (hunk 9 above)
Comment (YoGhurt111) [reply driver/chain_syncer/calldata/syncer.go:548]: Yes, i will add them

**Review by davidtaikocha — APPROVED** [review]

**Pull request merged**